blogs/WrappedYieldAwaitPattern.md

```diff
@@ -0,0 +1,103 @@
+Recently, I've been writing a lot of code which gets packaged up into libraries to be consumed by other developers. One of the key aspects of writing this sort of code effectively is to be stringent about argument validation. The earlier your code detects invalid arguments, the easier it is for consumers of your APIs to diagnose errors. Argument checking is fairly straight-forward, but I recently came upon an interesting subtlety when working with the C# <a href="http://msdn.microsoft.com/en-us/library/9k7k7cf0.aspx">yield</a> keyword. 
+
+For example, let's consider implementing the LINQ method <a href="http://msdn.microsoft.com/en-us/library/vstudio/bb534804(v=vs.100).aspx">Enumerable.TakeWhile</a>. TakeWhile has two simple conditions to check for: either argument can be null. We can easily implement TakeWhile using yield:
+
+<pre>
+public static IEnumerable<TSource> TakeWhile<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
+{
+    if (source == null) { throw new ArgumentNullException("source"); }
+	if (predicate == null) { throw new ArgumentNullException("predicate"); }
+	
+	foreach (var element in source) 
+	{
+		if (!predicate(element)) { break; }
+		yield return element; 
+	}
+}
+</pre>
+
+However, if we look at <a href="http://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,28936f7582207b50">Microsoft's implementation of TakeWhile</a>, we see something a bit different:
+
+<pre>
+public static IEnumerable<TSource> TakeWhile<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate) {
+	if (source == null) throw Error.ArgumentNull("source");
+	if (predicate == null) throw Error.ArgumentNull("predicate");
+	return TakeWhileIterator<TSource>(source, predicate);
+}
+
+static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate) {
+	foreach (TSource element in source) {
+		if (!predicate(element)) break;
+		yield return element;
+	}
+}
+</pre>
+
+This is what I call the <strong>Wrapped Yield/Await Pattern</strong>: the method is split into a public "wrapper" method that just does argument validation and then calls an inner private method that implements the actual iterator. Why is this done? Let's try running each implementation with invalid arguments:
+
+<pre>
+// MSFT's implementation: this line fails with ArgumentNullException
+var sequence = "abc".TakeWhile(default(Func<char, bool>));
+
+// our implementation: this line succeeds!
+var sequence = "abc".TakeWhile(default(Func<char, bool>));
```

Comment:
As a no longer C# developer it wasn't immediately clear to me that default(Func<char, bool>) was null.  It may be clearer (although less idiomatic) to write (Func<char, bool>)null.

But not a real issue.